src/java/org/apache/cassandra/index/sai/disk/v2/V2VectorIndexSearcher.java

```diff
@@ -254,38 +253,39 @@ private SparseFixedBitSet bitSetForSearch()
     }
 
     @Override
-    public RangeIterator<PrimaryKey> limitToTopResults(QueryContext context, RangeIterator<Long> iterator, Expression exp, int limit) throws IOException
+    public RangeIterator limitToTopResults(QueryContext context, List<PrimaryKey> keys, Expression exp, int limit) throws IOException
     {
         try (PrimaryKeyMap primaryKeyMap = primaryKeyMapFactory.newPerSSTablePrimaryKeyMap())
         {
-            // the iterator represents keys from all the segments in our sstable -- we'll only pull of those that
+            // the iterator represents keys from the whole table -- we'll only pull of those that
             // are from our own token range so we can use row ids to order the results by vector similarity.
             var maxSegmentRowId = metadata.toSegmentRowId(metadata.maxSSTableRowId);
             SparseFixedBitSet bits = bitSetForSearch();
             var rowIds = new IntArrayList();
             try (var ordinalsView = graph.getOrdinalsView())
             {
-                while (iterator.hasNext())
+                for (PrimaryKey primaryKey : keys)
                 {
-                    Long sstableRowId = iterator.peek();
+                    if (primaryKey.compareTo(metadata.minKey) < 0 || primaryKey.compareTo(metadata.maxKey) > 0)
+                        continue;
+                    // TODO this is O(log n), is this a hotspot?
+                    long sstableRowId = primaryKeyMap.rowIdFromPrimaryKey(primaryKey);
+                    // skip rows that are not in our segment (or more preciesely, have no vectors that were indexed)
+                    // or are not in this segment (which we can tell because the row id is Long.MAX_VALUE)
+                    if (sstableRowId < metadata.minSSTableRowId || sstableRowId == Long.MAX_VALUE)
+                        continue;
+
                     // if sstable row id has exceeded current ANN segment, stop
                     if (sstableRowId > metadata.maxSSTableRowId)
                         break;
 
-                    iterator.next();
-                    // skip rows that are not in our segment (or more preciesely, have no vectors that were indexed)
-                    if (sstableRowId < metadata.minSSTableRowId)
-                        continue;
-
                     int segmentRowId = metadata.toSegmentRowId(sstableRowId);
                     rowIds.add(segmentRowId);
-
+                    // TODO now that we know the size of keys evaluated, is it worth doing the brute
```

Comment:
it's important to have an accurate rows-involved count, so what we'd want to do (if we can't figure out a way to avoid hitting disk to look up row ids) is loop twice, one to get the set of keys in range, then check for brute force condition, then loop again to get rowids

Reply:
I agree that the double loop would solve the problem. I'll leave this open until we find a solution to the PK to row id question.